src/runtime_src/core/common/api/xrt_xclbin.cpp

```diff
@@ -38,126 +41,919 @@
+      m_axlf_sections[ASK_GROUP_CONNECTIVITY] = m_axlf_sections[CONNECTIVITY];
+    }
+  }
+
+  const std::vector<char>& get_data() const override
   {
     return m_axlf;
   }
+
+  std::pair<const char*, size_t>
+  get_axlf_section(axlf_section_kind section, const uuid& xclbin_id) const override
+  {
+    if (xclbin_id && xclbin_id != m_uuid)
+      throw std::runtime_error("xclbin id mismatch");
+    auto itr = m_axlf_sections.find(section);
+    return itr != m_axlf_sections.end()
+           ? (*itr).second
+           : std::make_pair(nullptr, size_t(0));
+  }
+
+protected:
+  std::vector<char> m_axlf;
+  std::map<axlf_section_kind, std::pair<const char*, size_t>> m_axlf_sections;
+
+  axlf*
+  get_top() const
+  {
+    return reinterpret_cast<axlf*>(const_cast<char*>(m_axlf.data()));
+  }
 };
 
-} //namespace
+/* class xclbin_filter - class for xclbin objects
+ *   This version stores only sections of interest
+ * Life time of xclbin are managed through shared pointers.
+ * A buffer is freed when last references is released.
+ */
+class xclbin_filter : public xclbin_impl
+{
+public:
+  explicit xclbin_filter(const axlf* top, const std::vector<axlf_section_kind>& kinds = ::xrt_core::xclbin_int::get_section_kinds())
+  {
+    if (strncmp(top->m_magic, "xclbin2", 7) != 0)
+      throw std::runtime_error("Invalid xclbin");
 
-////////////////////////////////////////////////////////////////
-// xrt_xclbin C++ API implementations (xrt_xclbin.h)
-////////////////////////////////////////////////////////////////
-namespace xrt {
+    m_uuid = uuid(top->m_header.uuid);
+
+    auto vbnv = reinterpret_cast<const char*>(top->m_header.m_platformVBNV);
+    m_vbnv = std::string(vbnv, strnlen(vbnv, 64));
+
+    m_axlf_sections.clear();
+
+    // Convert kinds to set
+    std::unordered_set<axlf_section_kind> kinds_set(kinds.begin(),kinds.end());
+
+    // Copy all section data of interest
+    for (const auto& section_header : xrt_core::xclbin_int::axlf_header_container(top)) {
+      auto kind = static_cast<axlf_section_kind>(section_header.m_sectionKind);
+      if (kinds_set.find(kind) != kinds_set.end()) {
+        auto section_data = reinterpret_cast<const char*>(top) + section_header.m_sectionOffset;
+        m_axlf_sections[kind] = std::vector<char>(section_data,section_data+section_header.m_sectionSize);
+      }
+    }
+
+    /* If group topology or group connectivity doesn't exists then use the
+     * mem topology or connectivity respectively section for the same.
+     */
+    if (m_axlf_sections.find(ASK_GROUP_TOPOLOGY) == m_axlf_sections.end() &&
+        m_axlf_sections.find(MEM_TOPOLOGY) != m_axlf_sections.end()) {
+      m_axlf_sections[ASK_GROUP_TOPOLOGY] = m_axlf_sections[MEM_TOPOLOGY];
+    }
+    if (m_axlf_sections.find(ASK_GROUP_CONNECTIVITY) == m_axlf_sections.end() &&
+        m_axlf_sections.find(CONNECTIVITY) != m_axlf_sections.end()) {
+      m_axlf_sections[ASK_GROUP_CONNECTIVITY] = m_axlf_sections[CONNECTIVITY];
```

Comment:
This logic has changed in XRT.  It is now possible to disable group sections.  Make sure you use xrt_core::xclbin::get_axlf_section(top, kind) to handle this logic.